src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,176 @@
+            f"`sha256` is the 64-character hex representation of the expected sha256 digest of the "
+            f"download file, as emitted by `shasum -a 256`, and `length` is the expected length of "
+            f"the download file in bytes. E.g., '3.1.2|darwin|"
+            f"6d0f18cd84b918c7b3edd0203e75569e0c7caecb1367bbbe409b44e28514f5be|42813'. "
+            f"Values are space-stripped, so pipes can be indented for readability if necessary."
+            f"You can compute the length and sha256 easily with:  "
+            f"curl -L $URL | tee >(wc -c) >(shasum -a 256) >/dev/null",
+        )
+        # For backwards compatibility with BinaryToolBase.  Subclasses of BinaryToolBase can
+        # instead subclass this class, and need only switch their --version-digest-mapping to
+        # --known-versions to get rid of the deprecation warning.
+        # Note: After removing this, also remove get_default_versions_and_digests() above.
+        register(
+            "--version-digest-mapping",
+            removal_version="1.30.0.dev2",
+            removal_hint="""Use --known-versions instead.""",
+            type=dict,
+            default=cls.get_default_versions_and_digests(),
+            fingerprint=True,
+            help="A dict mapping <platform constraint> -> (<version>, <fingerprint>, <size_bytes>)."
+            f'A "platform constraint" is any of {[c.value for c in PlatformConstraint]}, and '
+            "is the platform to fetch the tool for. A platform-independent tool should "
+            f"use {PlatformConstraint.none.value}, while a platform-dependent tool should specify "
+            'all environments it needs to be used for. The "fingerprint" and "size_bytes" '
+            "arguments are the result printed when running `sha256sum` and `wc -c` on "
+            "the downloaded file, respectively.",
+        )
+
+    @classmethod
+    @abstractmethod
+    def generate_url(cls, plat: Platform, version: str) -> str:
+        """Returns the URL for the given version of the tool, runnable on the given os+arch.
+
+        os and arch default to those of the current system.
+
+        Implementations should raise ExternalToolError if they cannot resolve the arguments
+        to a URL.  The raised exception need not have a message - a sensible one will be generated.
+        """
+
+    @classmethod
+    def generate_exe(cls, plat: Platform, version: str) -> str:
+        """Returns the archive path to the given version of the tool.
+
+        If the tool is downloaded directly, not in an archive, this can be left unimplemented.
+        """
+        return ""
+
+    @classmethod
+    def generate_request(
+        cls, plat: Platform, version: str, known_versions: List[str]
+    ) -> ExternalToolRequest:
+        for known_version in known_versions:
+            ver, plat_val, sha256, length = (x.strip() for x in known_version.split("|"))
```

Comment:
Consider validation for this failing here. Something like "Malformed string. Please run `./pants help-advanced {options-scope}` and read the documentation for `--known-versions`." 

You will want to ensure that the options scope is correctly presented with `-` instead of `_`.

Reply:
Good idea, will add validation.

Didn't get the comment about options scope?

Reply:
A good error message will include the full option name, including the subsystem’s option scope. You can’t know when writing this error message what that value will be because this class is meant to be subclassed and you expect the subclass to define the value. 

My warning is that you will need to accommodate for the risk of the subclass’s option scope having _ in its value, eg the options scope protoc_compiler. You will want to convert this into —protoc-compiler-known-versions, with a dash rather than an underscore.

Reply:
The options scope is simply `cls.options_scope`, no? And isn't that the scope we'll use in the help? And I don't think we need to start prefixing the flag in this message anyway. Users will get it.

Reply:
Indeed, it is `cls.options_scope`. No, it is not what we use in `./pants help`. Try `./pants help python_setup`; it fails. You need to use `./pants help python-setup`.

Reply:
Probably we should disallow underscores in scope names then, rather than fix them up on the fly.

Reply:
Or we should support either in help. Having to munge the scope here seems like the least ergonomic thing to do.